messages/proposed_event.go

```diff
@@ -1,14 +1,108 @@
 package messages
 
 import (
+	"encoding/json"
+
 	uuid "github.com/gofrs/uuid"
 )
 
 // ProposedEvent ...
 type ProposedEvent struct {
-	EventID      uuid.UUID
-	EventType    string
-	ContentType  string
-	Data         []byte
-	UserMetadata []byte
+	eventID      uuid.UUID
+	eventType    string
+	contentType  string
+	data         []byte
+	metadata []byte
+}
+
+func NewJsonProposedEvent(eventType string, payload interface{}) (ProposedEvent, error) {
+	event := ProposedEvent {}
+	bytes, err := json.Marshal(payload)
+
+	if err != nil {
+		return event, err
+	}
+
+	event.eventType = eventType
+	event.contentType = "application/json"
+	event.data = bytes
+	event.metadata = []byte{}
+
+	return event, nil
+}
+
+func NewBinaryProposedEvent(eventType string, bytes []byte) ProposedEvent {
+	event := ProposedEvent {
+		eventType: eventType,
+		contentType: "application/octet-stream",
+		data: bytes,
+		metadata: []byte{},
+	}
+
+	return event
+}
+
+func (event ProposedEvent) GetEventID() uuid.UUID {
+	return event.eventID
+}
+
+func (event ProposedEvent) GetEventType() string {
+	return event.eventType
+}
+
+func (event ProposedEvent) GetContentType() string {
+	return event.contentType
+}
+
+func (event ProposedEvent) GetData() []byte {
+	return event.data
+}
+
+func (event ProposedEvent) GetMetadata() []byte {
+	return event.metadata
+}
+
+func (event ProposedEvent) EventID(value uuid.UUID) ProposedEvent {
+	event.eventID = value
+	return event
+}
+
+func (event ProposedEvent) EventType(value string) ProposedEvent {
```

Comment:
Wouldn't it make API a bit nicer, if setters would modify `ProposedEvent` instead of returning a new one?

So instead of 

```go
event, err := messages.NewJsonProposedEvent("some-event", data)
// ...
event = event.ContentType("foo")
event = event.EventType("bar")
```
it will be just
```go
event, err := messages.NewJsonProposedEvent("some-event", data)
// ...
event.ContentType("foo")
event.EventType("bar")
```

I can't also imagine when immutability may be beneficial in that case.

The similar case is mentioned in https://golang.org/doc/effective_go#pointers_vs_values